client/client_test.go

```diff
@@ -610,3 +610,104 @@ func ExampleKV_RunTransaction() {
 	fmt.Println("Transaction example done.")
 	// Output: Transaction example done.
 }
+
+// concurrentConcatenation starts two Goroutines in parallel, both of which
+// read the other's key and try to append it on their own.
+// It is checked that the outcome is serializable, i.e. exactly one of the
+// two Goroutines (the later write) sees the previous write by the other.
+func concurrentConcatenation(kvClient *client.KV, t *testing.T) {
+	// wgStart waits for all transactions to line up, wgEnd has the main
+	// function wait for them to finish.
+	var wgStart, wgEnd sync.WaitGroup
+	wgStart.Add(2 + 1)
+	wgEnd.Add(2)
+
+	for i := 0; i < 2; i++ {
+		go func(i int) {
+			// Read the other key, write key i.
+			readKey := []byte(fmt.Sprintf("value-%d", (i+1)%2))
+			writeKey := []byte(fmt.Sprintf("value-%d", i))
+			value := []byte(fmt.Sprintf("value-%d", i))
+			defer wgEnd.Done()
+			wgStart.Done()
+			// Wait until the other goroutines are running.
+			wgStart.Wait()
+
+			txnOpts := &client.TransactionOptions{
+				Name: fmt.Sprintf("test-%d", i),
+			}
+			if err := kvClient.RunTransaction(txnOpts, func(txn *client.KV) error {
+				// Retrieve the other key.
+				gr := &proto.GetResponse{}
+				if err := txn.Call(proto.Get, proto.GetArgs(readKey), gr); err != nil {
+					return err
+				}
+
+				otherValue := []byte(nil)
+				if gr.Value != nil && gr.Value.Bytes != nil {
```

Comment:
Would be more understandable maybe to use Value.Integer instead of Value.Bytes for this. Then use addition instead of byte concatenation.